teps/0097-breakpoints-for-taskruns-and-pipelineruns.md

```diff
@@ -0,0 +1,317 @@
+
+Adding the name of the Task to `afterTask` spec will add a breakpoint after the last step of the TaskRun of the 
+Task mentioned in the spec.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: quarante-deux-deux
+spec:
+  debug:
+    breakpoint:
+      afterTask: ["deploy-frontend"]
+```
+
+### Breakpoint on failure of a TaskRun
+
+When we provide `onFailure` as `true` in the `breakpoint` spec of the PipelineRun, this will be applied to all the children TaskRun under
+this PipelineRun and hence `onFailure` will be present in the breakpoints of all the children TaskRuns.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: quarante-deux-deux
+spec:
+  debug:
+    breakpoint:
+      onFailure: true
+```
+
+## User Stories
+
+### CLI integration
+The scenario below provides details on what the user should do to debug their failed task.
+
+#### Environment access
+1. Get Failed TaskRun name.
+2. Provide as an argument to the cli to start the TaskRun in debug mode. To get
+   ```
+   tkn taskrun debug failed-taskrun-1234 --on-failure --console-access
+   ```
+   In the above example we are recreating the TaskRun with the following patch.
+   ```yaml
+   metadata:
+     name: failed-taskrun-1234-debug
+   spec:
+    debug: 
+      breakpoint: 
+        onFailure: true
+   ```
+3. Once the TaskRun is created and a step fails, the TaskRun Pod will be in the Running state. Till that time the CLI will wait 
+   for the TaskRun to stop executing (due to failure) and go into the limbo state which would be leveraged for debugging.
+   The CLI will open a shell to the step container which would be a reimplementation of `kubectl exec -it`.
+
+## Alternatives
+
+1. *"Breakpoint" as a map in the debugSpec.* (Used right now)
+
+    ```yaml
+    debug:
+      breakpoint: 
+        onFailure: true
+        beforeStep: []
+        afterStep: []
+    ```
+    The above shows what it could look like if the user could provide more points where they can halt the TaskRun i.e. before and after certain steps. 
+
+2. *"Breakpoint on failure" as a map.*
+
+    Based on 1. we can say that there might come a time, if rerunning steps does become a reality (which is highly likely), the user would like to debug before the failure takes place by enabling, something like beforeFailure which would rerun the step and halt it before execution.
+
+    ```yaml
+    debug:
+      breakpoint: 
+        onFailure: 
+          before: true
+          after: true
+        beforeStep: []
+        afterStep: []
+    ```
+
+3. *Enums rock* (Not applicable anymore)
+
+    Using a list instead of a map allow us to maintain the breakpoint locations in the Tekton API and not on the client side.
+
+    - Add a breakpoint before and after the failure.
+
+      ```yaml
+      debug:
+        breakpoint: ["beforeFailure", "onFailure"]
+      ```
+
+    - Create a breakpoint before and after execution of certain steps.
```

Comment:
please add the con about specializing strings - https://github.com/tektoncd/community/pull/572#pullrequestreview-836578705